src/cljc/athens/common_events/graph/ops.cljc

```diff
@@ -195,3 +196,39 @@
                            (map :block/new)
                            set)]
     new-uids))
+
+
+(defn structural-diff
+  "Calculates removed and added links (block refs & page links)"
+  [db ops]
+  (let [block-save-ops    (contains-op? ops :block/save)
+        new-blocks        (->> block-save-ops
+                               (map #(select-keys (:op/args %)
+                                                  [:block/uid :block/string])))
+        new-block-uids    (->> new-blocks
+                               (map :block/uid)
+                               set)
+        new-structures    (->> new-blocks
+                               (map :block/string)
+                               (map structure/structure-parser->ast))
+        old-block-strings (->> new-block-uids
+                               (map #(common-db/get-block-string db %)))
+        old-structures    (->> old-block-strings
+                               (map structure/structure-parser->ast))
+        links             (fn [structs names renames]
+                            (->> structs
+                                 (mapcat (partial tree-seq vector? identity))
+                                 (filter #(and (vector? %)
+                                               (contains? names (first %))))
+                                 (map #(vector (get renames (first %) (first %))
+                                               (-> % second :string)))
+                                 set))
+        old-links         (links old-structures
+                                 #{:page-link :hashtag :block-ref}
+                                 {:hashtag :page-link})
+        new-links         (links new-structures
+                                 #{:page-link :hashtag :block-ref}
+                                 {:hashtag :page-link})
+        removed-links     (set/difference old-links new-links)
+        added-links       (set/difference new-links old-links)]
+    [removed-links added-links]))
```

Comment:
Since this fn operates over atomic operations and is in the graph ops ns, it would be nice if the output uses identity in the sense defined in `athens.common-events.graph.schema`.

Right now it outputs what I think is parser names like `[:block-ref "123"]` and `[:page-link "123"]`, but in `athens.common-events.graph.schema` we have identity defined as `[:block/uid "123"]` and `[:page/title "123"]` instead.

Reply:
For now this data is used for reporting only, and it's ok this way.
We could change the returned format, but why bother if it's not used like that?